### ISSUE TITLE
Bug In esxi With ssh connection failure on first attempt, success on second

### ISSUE THREAD

**Issue by dsp9089** [josenk/vagrant-vmware-esxi#24]
Due to an esxi bug when AD authentication is enabled, the first ssh authentication attempt always fails, even when the password is correct.

As such, each attempt to connect via 'vagrant [command] --provider=vmware-esxi fails.

I unfortunately cannot disable AD authentication, as it's necessary in my environment.  Would it be possible to attempt authentication *twice* before raising the exception:
INFO interface: Machine: error-exit ["VagrantPlugins::ESXi::Errors::ESXiError", "There was an error talking to ESXi.\n  Unable to connect to ESXi host!"]

I realize that this is really an esxi issue, but I'm sure there are others who are in the same situation.

Thanks for your plugin.  I think it could make my life a *lot* easier.

**Comment by josenk** [comment]
Actually, you are the first to report this...  :-)  I don't have an esxi host connected with AD, so I'm unable to duplicate the issue....   If it's a known bug, do you have a link to show the issue?  bug id???

Have you tried this???   I have a feeling you might have an issue with your pam configuration...

https://www.linuxquestions.org/questions/linux-general-1/login-asks-for-password-twice-before-authenticating-169607/

**Comment by dsp9089** [comment]
Thanks for you response.

This does not appear to match the issue in the link you posted.  All the same, I did modify the pam configuration to use 'sufficient' rather than 'required' in the sshd configuration.  Unfortunately, that did not resolve the problem.

It's described pretty much as I see it here:
https://communities-gbot.vmware.com/message/2630250#2630250

with the only resolution suggested being to disable AD authentication.  

I did attempt to modify your code in lib/vagrant-vmware-esxi/action/esxi_password.rb as follows:
`Net::SSH.start(config.esxi_hostname, config.esxi_username,
                password:                   $esxi_password,
                port:                       config.esxi_hostport,
                keys:                       config.local_private_keys,
                timeout:                    20,
                number_of_password_prompts: 2,
                auth_methods:          password,
                non_interactive:            true`

However, this doesn't seem to make any difference whatever.  

In reading through the Net::SSH docs:
https://net-ssh.github.io/ssh/v1/chapter-2.html
http://www.rubydoc.info/github/net-ssh/net-ssh/Net/SSH

There doesn't seem to be a 'retries' parameter.
Looping (obviously) would be unhelpful, as the issue occurs during the Net::SSH.start() call.
I'm guessing that this needs to be resolved on the esxi side, and I'll start by upgrading to 6.5, as I'm currently on ESXi 6.0.0 Update 2.

As I initially said, this is an esxi bug and not an issue with your code.  I *really* appreciate your response and suggestions.  Thank you!

Unless you want to dig deeper into this (although I'm not sure why you would), this issue should be closed.

**Comment by josenk** [comment]
Have you tried using ssh keys?   If that works, you'll have to see how ovftool handles the AD bug???  Maybe you might get lucky with ovftool.

I also took a look at the net-ssh documentation and it's certainly not clear how number_of_password_prompts works when non_interactive = true...  Here is a very minimal ruby script to help you quickly test.  You can play around with some params to see if something works.

```
require 'net/ssh'

esxi_hostname = 'esxi'
esxi_username = 'root'
esxi_password = 'my_secret_password'
esxi_hostport = 22
local_private_keys = [ '~/.ssh/id_rsa', '~/.ssh/id_ecdsa', '~/.ssh/id_dsa' ]


begin
  Net::SSH.start(esxi_hostname, esxi_username,
                password:                   esxi_password,
                port:                       esxi_hostport,
                keys:                       local_private_keys,
                timeout:                    20,
                number_of_password_prompts: 1,
                non_interactive:            true
              ) do |ssh|

    esxi_version = ssh.exec!('vmware -v')
    ssh.close

    puts "ESXi version    : #{esxi_version}"
    exit 0
  end

rescue
  puts "There was an error connecting to esxi host"
end
```


If you still want to evaluate the plugin, you could always install esxi in a VM...  I haven't tried it, but it looks like it should be possible.  Hopefully you can disable AD authentication for POC/testing.

**Comment by dsp9089** [comment]
Thanks again for the suggestions.

As I said, the issue isn't your code, so testing with 'vagrant --debug up --provider=vmware-esxcli'  is plenty for testing, especially since I can just modify the Net::SSH.start call.

But the issue is now moot.  The reason I needed AD authentication was because of the VSphere thick (C#) client.  Now that I've upgraded to 6.5, the thick client is no longer supported -- obviating the need for AD auth.

Which actually works out, since the upgrade to 6.5 did *not* fix the two password prompt issue (grrr...maybe Xen or QEMU are in my future).  

Anyway, so I disabled AD authentication and am happily provisioning my first test VM with vagrant and your plugin as we speak.

Thanks again for your time and great suggestions!

Again, unless you have reason to keep it open, this issue should be closed.

Regards.
